### PR TITLE
clippy: Fix `vec_box` warnings in `components/script`

### PR DESCRIPTION
**Pull request by oluwatobiss**
Remove the unnecessary boxing of `Vec`'s content to fix the `vec_box` warnings.

**ref:** https://rust-lang.github.io/rust-clippy/master/index.html#vec_box

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #31500
- [X] These changes do not require tests because they only resolve clippy warnings. They do not change functionalities.

### CODE REVIEW

**Review by mrobinson — CHANGES_REQUESTED** [review]

File: components/script/dom/globalscope.rs [hunk 1]
```diff
@@ -294,15 +294,19 @@ pub struct GlobalScope {
...
 
     /// Promises in this list have previously been reported as rejected
     /// (because they were in the above list), but the rejection was handled
     /// in the last turn of the event loop.
     ///
     /// <https://html.spec.whatwg.org/multipage/#outstanding-rejected-promises-weak-set>
     #[ignore_malloc_size_of = "mozjs"]
-    consumed_rejections: DomRefCell<Vec<Heap<*mut JSObject>>>,
+    #[allow(clippy::vec_box)]
+    // `Heap` values must stay boxed, as they need semantics like `Pin` ie they cannot be moved
```
Comment (mrobinson) [inline components/script/dom/globalscope.rs:308]: I think it makes sense to put this above the `allow` directive here and everywhere.

File: components/script/dom/globalscope.rs [hunk 2]
```diff
@@ -2180,42 +2184,46 @@ impl GlobalScope {
     pub fn add_uncaught_rejection(&self, rejection: HandleObject) {
         self.uncaught_rejections
             .borrow_mut()
-            .push(*Heap::boxed(rejection.get()));
+            .push(Heap::boxed(rejection.get()));
```
Comment (mrobinson) [inline components/script/dom/globalscope.rs:2187]: Are these changes still necessary?

File: components/script/dom/globalscope.rs [hunk 3]
```diff
@@ -2180,42 +2184,46 @@ impl GlobalScope {
...
+            .position(|promise| *promise == Heap::boxed(rejection.get()))
         {
             uncaught_rejections.remove(index);
         }
     }
 
-    pub fn get_uncaught_rejections(&self) -> &DomRefCell<Vec<Heap<*mut JSObject>>> {
+    #[allow(clippy::vec_box)]
+    // `Heap` values must stay boxed, as they need semantics like `Pin` ie they cannot be moved
+    pub fn get_uncaught_rejections(&self) -> &DomRefCell<Vec<Box<Heap<*mut JSObject>>>> {
```
Comment (mrobinson) [inline components/script/dom/globalscope.rs:2203]: Please move the comment above the code it comments on.

File: components/script/dom/globalscope.rs [hunk 4]
```diff
@@ -2180,42 +2184,46 @@ impl GlobalScope {
...
         &self.uncaught_rejections
     }
 
     pub fn add_consumed_rejection(&self, rejection: HandleObject) {
         self.consumed_rejections
             .borrow_mut()
-            .push(*Heap::boxed(rejection.get()));
+            .push(Heap::boxed(rejection.get()));
```
Comment (mrobinson) [inline components/script/dom/globalscope.rs:2210]: Can this be reverted?

File: components/script/dom/globalscope.rs [hunk 5]
```diff
@@ -2180,42 +2184,46 @@ impl GlobalScope {
...
+            .position(|promise| *promise == Heap::boxed(rejection.get()))
         {
             consumed_rejections.remove(index);
         }
     }
 
-    pub fn get_consumed_rejections(&self) -> &DomRefCell<Vec<Heap<*mut JSObject>>> {
+    #[allow(clippy::vec_box)]
+    // `Heap` values must stay boxed, as they need semantics like `Pin` ie they cannot be moved
+    pub fn get_consumed_rejections(&self) -> &DomRefCell<Vec<Box<Heap<*mut JSObject>>>> {
```
Comment (mrobinson) [inline components/script/dom/globalscope.rs:2226]: Please move the comment above the code it comments on.

**Review by mrobinson — CHANGES_REQUESTED** [review]

File: components/script/dom/globalscope.rs (hunk 4 above)
Comment (oluwatobiss) [reply components/script/dom/globalscope.rs:2210]: Absolutely.

Comment (mrobinson) [reply components/script/dom/globalscope.rs:2210]: @sagudev ^^

Comment (sagudev) [reply components/script/dom/globalscope.rs:2210]: Am not sure I understand question. But given that we store Box<Heap<>> (and we must) we also need ` .push(Heap::boxed(rejection.get()))` (this is what we have now actually).

Comment (oluwatobiss) [reply components/script/dom/globalscope.rs:2210]: @sagudev To rephrase, since we are using the `allow` directive to permit a `<Vec<Box<Heap<*mut JSObject>>>>` syntax, would a `.push(Heap::boxed(rejection.get()))` be the correct code?

Or do we need to include a `*` (like so: `.push(*Heap::boxed(rejection.get()))`)?

Comment (mrobinson) [reply components/script/dom/globalscope.rs:2210]: Apologies, this looks fine. I was looking at the latest commit when I should have been looking at the diff from main. Please just move the comments. Otherwise this looks good.

File: components/script/dom/globalscope.rs [hunk 6]
```diff
@@ -294,6 +294,8 @@ pub struct GlobalScope {
     ///
     /// <https://html.spec.whatwg.org/multipage/#about-to-be-notified-rejected-promises-list>
     #[ignore_malloc_size_of = "mozjs"]
+    // `Heap` values must stay boxed, as they need semantics like `Pin` ie they cannot be moved
```
Comment (mrobinson) [inline components/script/dom/globalscope.rs:297]:
```suggestion
    // `Heap` values must stay boxed, as they need semantics like `Pin`
    // (that is, they cannot be moved).
```

File: components/script/dom/globalscope.rs [hunk 7]
```diff
@@ -302,6 +304,8 @@ pub struct GlobalScope {
     ///
     /// <https://html.spec.whatwg.org/multipage/#outstanding-rejected-promises-weak-set>
     #[ignore_malloc_size_of = "mozjs"]
+    // `Heap` values must stay boxed, as they need semantics like `Pin` ie they cannot be moved
```
Comment (mrobinson) [inline components/script/dom/globalscope.rs:307]:
```suggestion
    // `Heap` values must stay boxed, as they need semantics like `Pin`
    // (that is, they cannot be moved).
```

File: components/script/dom/globalscope.rs [hunk 8]
```diff
@@ -2194,6 +2198,8 @@ impl GlobalScope {
         }
     }
 
+    // `Heap` values must stay boxed, as they need semantics like `Pin` ie they cannot be moved
```
Comment (mrobinson) [inline components/script/dom/globalscope.rs:2201]:
```suggestion
    // `Heap` values must stay boxed, as they need semantics like `Pin`
    // (that is, they cannot be moved).
```

File: components/script/dom/globalscope.rs [hunk 9]
```diff
@@ -2215,6 +2221,8 @@ impl GlobalScope {
...
```
Comment (mrobinson) [inline components/script/dom/globalscope.rs:2224]:
```suggestion
    // `Heap` values must stay boxed, as they need semantics like `Pin`
    // (that is, they cannot be moved).
```

**Review by mrobinson — APPROVED** [review]
